packages/eslint-plugin/src/rules/prefer-promise-reject-errors.ts

```diff
@@ -0,0 +1,143 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import { getDeclaredVariables } from '@typescript-eslint/utils/eslint-utils';
+
+import {
+  createRule,
+  getParserServices,
+  isErrorLike,
+  isFunction,
+  isIdentifier,
+  isPromiseConstructorLike,
+  isPromiseLike,
+} from '../util';
+
+export type MessageIds = 'rejectAnError';
+
+export type Options = [
+  {
+    allowEmptyReject?: boolean;
+  },
+];
+
+export default createRule<Options, MessageIds>({
+  name: 'prefer-promise-reject-errors',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Require using Error objects as Promise rejection reasons',
+      recommended: 'strict',
+      extendsBaseRule: true,
+      requiresTypeChecking: true,
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          allowEmptyReject: {
+            type: 'boolean',
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+    messages: {
+      rejectAnError: 'Expected the Promise rejection reason to be an Error.',
+    },
+  },
+  defaultOptions: [
+    {
+      allowEmptyReject: false,
+    },
+  ],
+  create(context, [options]) {
+    const services = getParserServices(context);
+
+    function checkRejectCall(callExpression: TSESTree.CallExpression): void {
+      const argument = callExpression.arguments.at(0);
+      if (!argument && options.allowEmptyReject) {
+        return;
+      }
+
+      if (
+        !argument ||
+        !isErrorLike(services.program, services.getTypeAtLocation(argument))
+      ) {
+        context.report({
+          node: callExpression,
+          messageId: 'rejectAnError',
+        });
+      }
+    }
+
+    function skipChainExpression<T extends TSESTree.Node>(
+      node: T,
+    ): T | TSESTree.ChainElement {
+      return node.type === AST_NODE_TYPES.ChainExpression
+        ? node.expression
+        : node;
+    }
+
+    return {
+      CallExpression(node): void {
+        const callee = skipChainExpression(node.callee);
+
+        if (callee.type !== AST_NODE_TYPES.MemberExpression) {
+          return;
+        }
+
+        const calleeObjectType = services.getTypeAtLocation(callee.object);
+
+        const rejectMethodCalled = callee.computed
+          ? callee.property.type === AST_NODE_TYPES.Literal &&
+            callee.property.value === 'reject'
+          : callee.property.name === 'reject';
+        const calleeIsLikePromise =
+          isPromiseConstructorLike(services.program, calleeObjectType) ||
+          isPromiseLike(services.program, calleeObjectType);
+
+        if (!rejectMethodCalled || !calleeIsLikePromise) {
```

Comment:
[Performance] Previously, when it was all in one `if` statement, the type-checking functions `isPromiseConstructorLike` and `isPromiseLike` wouldn't be called if the previous logical bits (what is now `rejectMethodCalled`) were falsy.

So now it's a bit more readable at the cost of performance. One step forward, one step back!

My intent was to suggest splitting into functions so the logic gets lazily evaluated:

```ts
if (!rejectMethodCalled(callee) || !calleeIsLikePromise(callee)
```

...where `calleeObjectType` is only created inside `calleeIsLikePromise`. That way we don't call to the type checker at all if we don't need. The type checker can be slow sometimes so this is very good.

Aside: I filed an issue on myself to find a lint rule to enforce this: https://github.com/JoshuaKGoldberg/create-typescript-app/issues/1188

Reply:
Ohh, don't know how I didn't notice that, thanks! Also thx for detailed explanation, I changed it https://github.com/typescript-eslint/typescript-eslint/pull/8011/commits/f186084686347b0f3850bf8cbb1ad07818e267c2